netket/experimental/driver/vmc_kernel.py

```diff
@@ -0,0 +1,164 @@
+    else:
+        shape = jnp.zeros((int(matrix_side/n_nodes),), dtype=jnp.float64)
+        aus_vector, token = mpi_scatter_jax(shape, token=token)
+
+    updates = O_L.T @ aus_vector
+    updates, token = mpi_allreduce_sum_jax(updates, token=token)
+    
+    return -updates 
+
+inv_default_solver = lambda A, b: jnp.linalg.inv(A) @ b
+linear_solver = lambda A, b: jsp.linalg.solve(A, b, assume_a="pos")
+
+class VMC_kernelSR(VMC):
+    """
+    Energy minimization using Variational Monte Carlo (VMC) and the kernel 
+    formulation of Stochastic Reconfiguration (SR). This approach lead to
+    *exactly* the same parameter updates of the standard SR with a
+    diagonal shift regularization. For this reason, it is equivalent to the standard 
+    nk.driver.VMC with the preconditioner nk.optimizer.SR(solver=netket.optimizer.solver.solvers.solve))
+    In the kernel SR framework, the updates of the parameters can be written as:
+
+    :math:`\delta \\theta = \\tau X(X^TX + \lambda \mathbb{I}_{2M})^{-1} f`,
+
+    where :math:`X \in R^{P \\times 2M}` is the concatenation of the real and imaginary part 
+    of the centered Jacobian, with P the number of parameters and M the number of samples.
+    The vector f is the concatenation of the real and imaginary part of the centered local
+    energy. Note that, to compute the updates, it is sufficient to invert an :math:`M\\times M` matrix
+    instead of a :math:`P\\times P` one. As a consequence, this formulation is useful
+    in the typical deep learning regime where :math:`P \gg M`.
+    
+    See R.Rende, L.L.Viteritti, L.Bardone, F.Becca and S.Goldt (https://arxiv.org/abs/2310.05715)
+    for a detailed description of the derivation. A similar result can be obtained by minimizing the
+    Fubini-Study distance with a specific constrain, see A.Chen and M.Heyl (https://arxiv.org/abs/2302.01941) 
+    for details.
+    """
+
+    def __init__(
+        self,
+        hamiltonian: AbstractOperator,
+        optimizer,
+        diag_shift,
+        *args,
+        linear_solver_fn: Callable[[jax.Array, jax.Array], jax.Array] = linear_solver,
+        jacobian_mode = None,
+        variational_state = None,
+        **kwargs,
+    ):
+        """
+        Initializes the driver class.
+
+        Args:
+            hamiltonian: The Hamiltonian of the system.
+            optimizer: Determines how optimization steps are performed given the
+                    bare energy gradient.
+            diag_shift: The diagonal shift of the stochastic reconfiguration matrix.
+                    Typical values are 1e-4 ÷ 1e-3.
+        """
+        super().__init__(hamiltonian, optimizer, variational_state=variational_state)
+
+        self._ham = hamiltonian.collect()  # type: AbstractOperator
+        _, self.unravel_params_fn = ravel_pytree(self.state.parameters)
+        self.diag_shift = diag_shift
+        self.jacobian_mode = jacobian_mode
+        self.linear_solver_fn = linear_solver_fn
```

Comment:
let's mark this internal for now. 
```suggestion
        self._linear_solver_fn = linear_solver_fn
```